### PR TITLE
add dark mode toggle if dark.css is loaded

### LINKED ISSUE #241: Dark Mode
There's been two PRs so far that tried to add Dark Mode support, https://github.com/w3c/tr-design/pull/184 from @jakubklimek and https://github.com/w3c/tr-design/pull/202 from @tabatkins but we haven't merged any of them into the main line because swapping a spec that isn't expecting it into dark mode can make things unreadable.

Originally we were thinking to plan out the addition of Dark Mode with plenty of advance warning so that editors would have time to fix their specs before we deployed it across all of w3.org/TR, and to only apply it to specs published after that deploy date, with the hope that all the specs that got published from that point in time forward would be compatible with Dark Mode... and were waiting for a good time to prepare for and apply that cut-off.

But now I'm thinking what might make more sense is to add a separate dark.css file which editors can selectively choose to add to their spec, presumably after verifying that their diagrams and any custom styling are compatible with it. We could deploy that basically immediately, and the editors can opt in whenever they're ready.

Relatedly, we probably want a way for users to flip the colors on/off, and to have that persist across w3.org/TR so that users can opt into or out of dark mode for our specs generally. (Certain vendors seem to conflate the idea of “I want a dark OS UI” with “I want all content on the Web to be dark”...) I'm not sure how to implement that, though.

### PR DESCRIPTION
**Pull request by deniak**
This is an attempt to add a button to switch between dark/light mode.
Example can be seen [here](https://cdn.statically.io/gh/w3c/tr-design/dark-mode-toggle/src/README.html)

I'm not sure where the button should be placed so I made it sticky at the bottom of the page but feel free to edit the PR if there's a better position.

Fix #241

### CODE REVIEW

File: src/fixup.js
Comment (marcoscaceres) [inline src/fixup.js]: Adding directly to the body doesn't feel right. I wonder if we should start some kind of `menu` or something..

File: src/fixup.js
Comment (marcoscaceres) [inline src/fixup.js]: As we are never really sure what we are going to get from user land code... 

```suggestion
    const currentTheme = localStorage.getItem("tr-theme") || "light";
    switch (currentTheme) {
      case "light":
         darkCss.disabled = true;
      case "dark":
      default:
         darkCss.disabled = false;
    }
```

**Review by marcoscaceres — COMMENTED** [review]
Couple of suggestions

File: src/fixup.js
Comment (deniak) [reply src/fixup.js]: yeah, I couldn't decide where that button should go. @fantasai, any suggestion on that button? The style could definitely be improved as well.

Comment (marcoscaceres) [reply src/fixup.js]: Looking at how others do this, I think we need a UI component that can be in one of 3 states... "light", "dark", "auto". 

The best I've seen is how Apple does it (of course): 

Screen Shot 2021-06-08 at 1 03 40 pm

They use a `<input type=radio>`, which I think makes sense. 

It also means we don't need to disable the stylesheet. We can just include it as "auto", and then let the use decide which mode they explicitly prefer.

Comment (marcoscaceres) [reply src/fixup.js]: If you've like to see how it works at Apple, scroll to the bottom of https://developer.apple.com

File: src/fixup.js
Comment (marcoscaceres) [inline src/fixup.js:302]: Probably want to check for support... something like: 
```suggestion
  if (darkCss && matchMedia("(prefers-color-scheme)").matches) {
```

File: src/fixup.js
Comment (deniak) [reply src/fixup.js]: Thanks @marcoscaceres, I'm actually not a big fan of disabling the stylesheet also but I wasn't sure if "duplicating" the style between the different `prefers-color-scheme` was OK, which, AFAIK, is the only alternative to switch between the schemes.
I've just submitted a version using a `body` data attribute to specify the scheme. I also tried to reproduce the style of the buttons from Apple.

File: src/fixup.js
Comment (deniak) [reply src/fixup.js:302]: Suggestion applied in https://github.com/w3c/tr-design/pull/243/commits/2344038765001a652794626b7d41b70c69779e9e

File: src/fixup.js
Comment (fantasai) [reply src/fixup.js]: I think disabling the style sheet or modifying its 'media' attribute is a perfectly reasonable way to handle this, fwiw. Why wouldn't it be?

File: src/fixup.js
Comment (marcoscaceres) [inline src/fixup.js]: React-hipster-wanna-be-version: 

```suggestion
  /* Dark mode toggle */
  const darkCss = document.querySelector('link[href="dark.css"]');
  if (darkCss && matchMedia("(prefers-color-scheme)").matches) {
    const render = document.createElement("x-temp");
    function createOption(option) {
      const checked = option === document.body.dataset.colorScheme;
      return `
        <label>
          <input name="color-scheme" type="radio" value="${option}" checked="${checked}">
          <span>${option}</span>
        </label>
      `.trim();
    }
    render.innerHTML = `
      <div id="theme-toggle" role="radiogroup" aria-label="Select a color scheme">
        ${["light", "dark", "auto"].map(createOption).join("")}
      </div>
    `;
    const changeListener = (event) => {
      const { value } = event.target;
      document.body.dataset.colorScheme = value;
      localStorage.setItem("tr-theme", value);
    };
    render.querySelectorAll("input[type='radio']").forEach((input) => {
      input.addEventListener("change", changeListener);
    });
    document.querySelector("div.head").append(...render.children);
  }
```

Comment (deniak) [reply src/fixup.js]: Thanks for the suggestion. All looks great except that:

1. it's missing the local storage check at the beginning `document.body.dataset.colorScheme = localStorage.getItem("tr-theme") || "auto";`
2. the `checked` attribute l.308 should only be present for one and only one input.

Comment (deniak) [reply src/fixup.js]: @marcoscaceres I've included your suggesting in https://github.com/w3c/tr-design/pull/243/commits/101f56929151e563ee39cc8edabbf13d8d1a413d

File: src/fixup.js
Comment (deniak) [reply src/fixup.js]: I don't have a strong opinion on disabling a stylesheet. I don't recall seeing this somewhere else so it seems a bit strange to me but I don't inspect all the pages I visit :)
I've submitted a [new version](https://github.com/w3c/tr-design/pull/243/commits/101f56929151e563ee39cc8edabbf13d8d1a413d) which disables the dark stylesheet in light mode is selected. It also includes @marcoscaceres refactoring: see https://cdn.statically.io/gh/w3c/tr-design/dark-mode-toggle/src/README.html

Comment (marcoscaceres) [reply src/fixup.js]: Looking pretty cool! We probably want to set the background though:   
image

I'm also a bit unsure about having it persistently on screen. 

Trying to figure out where it could live... maybe, on mobile, it should be part of the ToC sidebar?

**Comment by fantasai** [comment]
Pushed dark.css to the repo in https://github.com/w3c/tr-design/commit/7989a2e622c9d37a2ef9c01856dbcfcd271b9d04 using the colors from Bikeshed (plus an additional set for the amendment boxes, which Bikeshed didn't have). README uploaded at https://fantasai.inkedblade.net/style/design/w3c-restyle/2021/README

CC @tabatkins

**Comment by deniak** [comment]
I've updated the PR to use the stylesheet @fantasai just pushed. `fixup.js` now tweaks the `media` attribute of the stylesheet if the user tries to force either the light or dark mode.

@fantasai, do you have any preference on the location of the buttons? As @marcoscaceres suggested, maybe next to the `<h2>Table of Contents</h2>`?

File: src/base.css
Comment (fantasai) [inline src/base.css]: I think we should not add this. Currently we have a transparent background, and probably people are depending on that in places.

Note: dark.css has such a rule for safety in dark mode... But since we are opt-in rather than by default like Bikeshed, we could also consider removing it and letting editors evaluate their images as needed. Something to discuss...

File: src/fixup.js
Comment (fantasai) [inline src/fixup.js]: This needs to be $= otherwise it'll never match except for the README!

File: src/fixup.js
Comment (fantasai) [inline src/fixup.js]: I think you want to remove the `matchMedia("(prefers-color-scheme)").matches`. If someone doesn't have a preferred default color scheme, seems OK for them to still get a toggle!

File: src/fixup.js
Comment (fantasai) [inline src/fixup.js]: why `x-temp` rather than `div`?

**Review by fantasai — CHANGES_REQUESTED** [review]

File: src/fixup.js
Comment (annsalvaje) [reply src/fixup.js:306]: Fsmhjh

Comment (marcoscaceres) [reply src/fixup.js]: No reason... I just like using generic custom element, but a `div` would work just as well.

File: src/fixup.js
Comment (marcoscaceres) [reply src/fixup.js]: agree

File: src/fixup.js
Comment (marcoscaceres) [reply src/fixup.js]: oops, yes.

File: src/fixup.js
Comment (fantasai) [reply src/fixup.js]: @marcoscaceres I feel like div is better than "generic custom element"... If you're going to make a custom element, at least have a decent justification for it!

File: src/fixup.js
Comment (deniak) [reply src/fixup.js]: addressed: https://github.com/w3c/tr-design/pull/243/commits/9efe263c396ab35f998f9d313532176e6851367e

File: src/fixup.js
Comment (deniak) [reply src/fixup.js]: addressed: https://github.com/w3c/tr-design/pull/243/commits/9efe263c396ab35f998f9d313532176e6851367e

File: src/fixup.js
Comment (deniak) [reply src/fixup.js]: addressed: https://github.com/w3c/tr-design/pull/243/commits/9efe263c396ab35f998f9d313532176e6851367e

File: src/fixup.js
Comment (deniak) [reply src/fixup.js]: Note that I would expect the link to be something like https://www.w3.org/StyleSheets/TR/2016/dark.css once the PR is merged

**Comment by deniak** [comment]
I've just pushed some of the changes we discussed. You can check the result [here](https://raw.githack.com/w3c/tr-design/dark-mode-toggle/src/README.html).

@fantasai any chance you can ask mpt to send me the logo he drew? I only have a screenshot of the logo.

**Review by fantasai — APPROVED** [review]
Please use the Squash-Merge option when merging.

**Comment by deniak** [comment]
@fantasai @marcoscaceres, thanks for the review.
We might have a small remaining issue because of that rule: https://www.w3.org/pubrules/doc/rules/?profile=WD#lastStylesheet

That PR works only if `dark.css` is declared *after* the `W3C-*.css` stylesheet but pubrules currently checks that the latter is the last stylesheet.
I'll merge that PR after adding the exception to pubrules.

**Comment by deniak** [comment]
I just updated the PR to specify the use of `https://www.w3.org/StyleSheets/TR/2016/dark.css` instead of any dark stylesheet.

**Comment by deniak** [comment]
Just released the new version of pubrules to allow dark.css.

**Comment by deniak** [comment]
The script and stylesheets have been deployed. You may now include https://www.w3.org/StyleSheets/TR/2016/dark.css *after* the W3C stylesheet to enable the dark mode!

**Pull request merged**
